File: Challenge-004-Dynamic-Stack/java/Stack.java
public class Stack {
    private static final int MAX_SIZE = 100;
    private int[] items = new int[MAX_SIZE];
    private int top = -1;

    public boolean isEmpty() {
        return top == -1;
    }

    public boolean isFull() {
        return top == MAX_SIZE - 1;
    }

    public void push(int value) {
        if (isFull()) {
            System.out.println("Stack overflow");
            return;
        }
        items[++top] = value;
    }

    public void pop() {
        if (isEmpty()) {
            System.out.println("Stack underflow");
            return;
        }
        top--;
    }

    public int peek() {
        if (isEmpty()) {
            System.out.println("Stack is empty");
            return -1;
        }
        return items[top];
    }
}

File: Challenge-004-Dynamic-Stack/JavaScript/Stack.js
class Stack {
    constructor() {
        this.items = [];
    }

    // Agregar elemento (push)
    push(element) {
        this.items.push(element);
    }

    // Eliminar elemento (pop)
    pop() {
        if (this.isEmpty()) {
            console.log("Stack underflow");
            return undefined;
        }
        return this.items.pop();
    }

    // Consultar elemento en la cima (peek)
    peek() {
        if (this.isEmpty()) {
            console.log("Stack is empty");
            return undefined;
        }
        return this.items[this.items.length - 1];
    }

    // Verificar si está vacía
    isEmpty() {
        return this.items.length === 0;
    }
}

// Ejemplo de uso
const stack = new Stack();
stack.push(10);
stack.push(20);
stack.push(30);

console.log("Cima:", stack.peek()); // 30
console.log("Pop:", stack.pop());   // 30
console.log("Nueva cima:", stack.peek()); // 20